ReportCalculator.cs

```diff
@@ -6,36 +6,48 @@ public class ReportCalculator
         public Discount discount { get; set; }
         public UPCDiscount upcDiscount { get; set; }
```

Comment:
What about moving all the discounts under a parent type, collecting them all in a single collection, and then using polymorphism to distinguish between types?